### PR TITLE
Restore definitions for .get and .join

### PR DESCRIPTION
**Pull request by demurgos**
Issue #5

### CODE REVIEW

File: bluebird.d.ts [hunk 1]
```diff
@@ -151,8 +151,8 @@ declare class Bluebird<R> implements Bluebird.Thenable<R>, Bluebird.Inspection<R
    * });
    * </code>
    */
-  // TODO find way to fix get()
-  // get<U>(propertyName: string): Bluebird<U>;
+  // TODO: explain original comment from DT: TODO find way to fix get()
```
Comment (louy) [inline bluebird.d.ts]: The point was to use something like subtypes (in TS roadmap I think) to get the type `U` automatically.

File: bluebird.d.ts [hunk 2]
```diff
@@ -459,8 +459,22 @@ declare class Bluebird<R> implements Bluebird.Thenable<R>, Bluebird.Inspection<R
...
+  static join<R, A1, A2>(arg1: A1 | Bluebird.Thenable<A1>, arg2: A2 | Bluebird.Thenable<A2>, handler: (arg1?: A1, arg2?: A2) => R | Bluebird.Thenable<R>): Bluebird<R>;
+  static join<R, A1, A2, A3>(arg1: A1 | Bluebird.Thenable<A1>, arg2: A2 | Bluebird.Thenable<A2>, arg3: A3 | Bluebird.Thenable<A3>, handler: (arg1?: A1, arg2?: A2, arg3?: A3) => R | Bluebird.Thenable<R>): Bluebird<R>;
+  static join<R, A1, A2, A3, A4>(arg1: A1 | Bluebird.Thenable<A1>, arg2: A2 | Bluebird.Thenable<A2>, arg3: A3 | Bluebird.Thenable<A3>, arg4: A4 | Bluebird.Thenable<A4>, handler: (arg1?: A1, arg2?: A2, arg3?: A3, arg4?: A4) => R | Bluebird.Thenable<R>): Bluebird<R>;
+  static join<R, A1, A2, A3, A4, A5>(arg1: A1 | Bluebird.Thenable<A1>, arg2: A2 | Bluebird.Thenable<A2>, arg3: A3 | Bluebird.Thenable<A3>, arg4: A4 | Bluebird.Thenable<A4>, arg5: A5 | Bluebird.Thenable<A5>, handler: (arg1?: A1, arg2?: A2, arg3?: A3, arg4?: A4, arg5?: A5) => R | Bluebird.Thenable<R>): Bluebird<R>;
+
   // variadic array with promises of value
   static join<R>(...values: Bluebird.Thenable<R>[]): Bluebird<R[]>;
```
Comment (louy) [inline bluebird.d.ts:479]: I think this signature should be removed.

Comment (demurgos) [inline bluebird.d.ts:479]: I think that the whole point of `join` is to easily resolve a few distinct promises and then handle them. The current two declarations do not cover this use and cause warnings for some valid code.

From [the documentation](http://bluebirdjs.com/docs/api/promise.join.html):

``` javascript
var Promise = require("bluebird");
var join = Promise.join;

join(getPictures(), getComments(), getTweets(),
    function(pictures, comments, tweets) {
    console.log("in total: " + pictures.length + comments.length + tweets.length);
});
```

Which signature would you use to describe this behaviour ? This is a valid way to use this method so it should be declared.

File: bluebird.d.ts (hunk 1 above)
Comment (demurgos) [inline bluebird.d.ts]: Ok, thank you for your explanation. Are these subtypes ready yet ?
Anyway, it's better to use a sub-optimal declaration than no declaration at all.

File: bluebird.d.ts (hunk 2 above)
Comment (louy) [inline bluebird.d.ts:479]: I mean the last two. the ones without a handler as the last argument.

File: bluebird.d.ts (hunk 1 above)
Comment (louy) [inline bluebird.d.ts]: I think they'll be in 2.0 or something like that. agreed we shouldn't wait for them. I was just explaining the TODO comment.

File: bluebird.d.ts (hunk 2 above)
Comment (demurgos) [inline bluebird.d.ts:479]: Ok sorry (Github displayed the part with the handler).
I agree with you about this, I just did not want to break anything.

File: bluebird.d.ts (hunk 1 above)
Comment (demurgos) [inline bluebird.d.ts]: Thank you, I'll edit the comment with your explanation.

Comment (demurgos) [inline bluebird.d.ts]: I just checked the TS roadmap and couldn't find any mention of subtypes. Could you clarify what you mean ?

**Comment by unional** [comment]
Thanks

**Comment by demurgos** [comment]
Thank you for the merge. Can you register this version ?

**Comment by unional** [comment]
Are they part of 3.4.0?  May need to branch and add them in too

**Comment by demurgos** [comment]
`Promise.join` and `.get` are part of bluebird since a long time (at least 1.x for join, not sure for .get but I already used it with 2.x; 3.4.0 still has them) and this PR just fixed some incorrect declarations. If you do have multiple declarations for multiple versions of bluebird then yes, I would advise to add this fix everywhere.

**Comment by unional** [comment]
Ok.

**Comment by demurgos** [comment]
Thank you for replying quickly. I was able to download the latest version from the registry and my projects compile again without warnings.

File: bluebird.d.ts (hunk 2 above)
Comment (unional) [inline bluebird.d.ts:479]: @louy, exactly which signature you refer to? The `(...values)` ones?

Comment (unional) [inline bluebird.d.ts:479]: based on https://github.com/petkaantonov/bluebird/blob/master/src/join.js#L100
It is supported

Comment (unional) [inline bluebird.d.ts:479]: Although I wonder should it be `static join<R>(...values: Array<R | Thenable>)...`

Comment (louy) [inline bluebird.d.ts:479]: @unional this: https://github.com/Microsoft/TypeScript/issues/1295
sorry I was mistaken, it's not on the roadmap but they're accepting PRs.

**Pull request merged**